storage/innobase/row/row0ins.cc

```diff
@@ -2658,6 +2657,41 @@ row_ins_clust_index_entry_low(
 				goto skip_bulk_insert;
 			}
 
+#ifdef WITH_WSREP
+			if (wsrep_thd_is_local(trx->mysql_thd))
+			{
+				char	db_buf[NAME_LEN + 1];
+				char	tbl_buf[NAME_LEN + 1];
+				ulint	db_buf_len, tbl_buf_len;
+
+				if (!index->table->parse_name(db_buf, tbl_buf, &db_buf_len, &tbl_buf_len))
+				{
+					trx->error_state = DB_ROLLBACK;
+					goto commit_exit;
+				}
+
+				/* Append table-level exclusive key for bulk insert. */
```

Comment:
Could this be replaced by wsrep service call `wsrep_thd_append_table_key(trx->mysql_thd, db, table, WSREP_SERVICE_KEY_EXCLUSIVE)`? This way the `wsrep_prepare_keys_for_isolation()/wsrep_keys_free()` would not have to be exposed on InnoDB side, making InnoDB patch here smaller.

Reply:
Do I understand it correctly that you are proposing to introduce a new service call `wsrep_thd_append_table_key()` instead of exposing the `wsrep_prepare_keys_for_isolation()/wsrep_keys_free()` pair?

Reply:
Or even better, can `wsrep_thd_append_key()` used?

Reply:
If we have '\0' terminated `db` and `table` strings and we are always appending table level key, allocating key parts for key array seems excessive (and complicates things unnecessarily, the return value of `wsrep_prepare_keys_for_isolation()` should also be checked here). Having single call which would convert db/table into wsrep representation internally would be much simpler.

Reply:
Okay